### ISSUE TITLE
Capture environment and code identity in training run artifacts

### ISSUE THREAD

**Issue by magilliam27** [magilliam27/MCI-GRU#144]
**Execution gate satisfied — 2026-09-12:** PR #204 merged with explicit owner approval at `f81573a14a35cdc53975640c28df8d3fc39b5756`; #142 is closed. Historical native dependency is retained, with no open blocker. Current merged target and input-preservation contracts govern execution.

**Mode:** AFK implementation under the Pocock workflow, with genuinely unconfirmed public test boundaries routed to the owner through PM.

**Dispatch status — 2026-09-13:** startup verified for the owner-approved GPT-6 Astra / Extra High worker, task `01a098ee-6d58-7671-8094-24992dfc3a1c`, in isolated worktree `C:/Users/magil/.codex/worktrees/39b9/MCI-GRU`. Freshly fetched `origin/main` is `f81573a14a35cdc53975640c28df8d3fc39b5756`; this clean checkout was aligned to it before branching. Assigned to `magilliam27`; exact ownership below is unchanged. Bounded read-only preparation is active; the first public capture-to-persisted-read-back seam is awaiting owner sequencing through PM, and no tests or implementation have started. PM task: `01a097d2-d237-7af2-b389-cd34409125d0`.

**Owned paths:** `mci_gru/evaluation/execution_provenance.py` (new); `mci_gru/evaluation/selection_audit.py` only shared code-identity/version helper extraction with compatibility; `run_experiment.py` only execution-start/per-window provenance linkage; `mci_gru/evaluation/experiment_summary.py` only resolved-config linkage and captured execution-record serialization; `mci_gru/training/ensemble.py` only actual member seed/device event reporting; `mci_gru/training/trainer.py` only effective precision/backend reporting; `tests/test_execution_provenance.py` and `tests/test_execution_provenance_integration.py` (new); `docs/agents/guide.md`; coordinated regeneration of `docs/TEST_REGISTRY.md`. No training behavior change. Exclude #191's observed-input collector/fingerprints, #206 loaders/snapshots, #208 attachment/bundle implementation, and #205 hashing ownership. Shared source hunks require coordination before editing; extend any additional scope through PM first.

Part of #140

**Prior documentation gate:** the reproducibility-target child (D1) is now closed; the merged target contract defines what must be captured.

## Measured state

Published `run_metadata.json` records exactly: `norm_means`, `norm_stds`, `feature_cols`,
`kdcode_list`, `his_t`, `label_t`, `seed`, `train_end`, `data_file`.

Checked on both the 2025 and the 2026 published results:

- no interpreter or package versions
- no git commit
- no `data_file_sha256`
- `data_file` is a bare path; on the 2026 result it is an absolute `C:\Users\magil\...` path

So no published result can be reproduced, and the reasons stack at four independent layers --
environment, code identity, data, and training determinism (#132). Any one of them alone is
sufficient to block it.

## The pattern already exists

`mci_gru/evaluation/selection_audit.py:721-761` already emits
`mci_gru.selection_research_code_identity.v1`:

```python
return {
    "schema": "mci_gru.selection_research_code_identity.v1",
    "git_commit": git_commit,
    "git_dirty_diff_sha256": hashlib.sha256(diff).hexdigest(),
    "working_tree_source_sha256": source_tree_hash,
    "source_hashes": source_hashes,
    "runtime_versions": {
        "numpy": ..., "pandas": ..., "python": ..., "scipy": ...,
    },
}
```

It is called only from the selection-research protocol, never from the training path. Two adjacent
mechanisms also already exist: `write_resolved_config` emits `resolved_config_sha256`, and
`data_file_fingerprint` emits `data_file_sha256` and is wired at `run_experiment.py:194`. The
published `seed_results/` predate the latter.

This ticket propagates and extends a tested pattern. It does not design a provenance system.

## Work

1. Emit code identity on the training path, reusing the existing schema rather than inventing one.
2. **Add `torch` and `torch-geometric` to `runtime_versions`.** Currently numpy, pandas, python, and
   scipy only -- omitting the two packages that differ most across the three known stacks
   (torch 2.11.0 / 2.12.1 / 2.13.0) and whose numerics are most likely to move.
3. Confirm `data_file_fingerprint` reaches every run-artifact path, including the notebook-driven
   runs that produced `seed_results/`.
4. Capture whatever D1 additionally requires -- for the environment-equivalent target that means the
   device and the cuDNN / deterministic-algorithms state, per #132's option 1.

## Evidence requirement

Load-bearing, and non-negotiable per repository standard: for each new provenance field, **break the
capture, confirm the guarding test fails, restore, confirm it passes, and report the table.** Three
shipped defects in this repository survived because their guarding tests were vacuous.

Any branch adding or renaming a test must regenerate `docs/TEST_REGISTRY.md` or CI lint fails; run
with `--junitxml=test_reports\junit.xml` first so statuses are recorded rather than blank.

## Acceptance criteria

- [ ] Training runs emit code identity and runtime versions including torch.
- [ ] `data_file_sha256` present on every run-artifact path.
- [ ] Mutation-check table reported for each new field.
- [ ] `docs/TEST_REGISTRY.md` regenerated if tests were added or renamed.

## Noticed, not authorized by this ticket

The 2026 published result's model artifacts are not in the repository.
`seed_results/2026/seed73_dynamic_m/seed73_dynamic_m.lnk` is a git-tracked Windows shortcut to
`C:\Users\magil\Downloads\mci_gru_all_results (6)\...`, sitting beside seven sibling copies whose
distinguishing suffix is a browser download counter. The target currently exists and holds
`graph_data.pt` and ten `model_N_best.pth` checkpoints; `*.pt` and `*.pth` are gitignored, so that
is the entire artifact chain. Deciding what to do about it is an owner call.

## Current execution-time acceptance refinement

Capture effective resolved config, actually applied member seeds, recoverable source state and observed environment at execution time, for stock/index and each walk-forward window, including incomplete attempts. Preserve existing metadata and emit references #208 can attach without reobserving the current machine. Planned seeds are not proof that members ran. Include Python/packages (torch and torch-geometric), effective device, CUDA/driver/cuDNN, determinism/precision flags, observed memory/RAM and Colab label where actually observed; unavailable fields remain explicit. A dirty source hash must resolve to retained recoverable source bytes, including relevant untracked source; credentials are excluded. Preserve existing redacted config artifacts and bind effective input settings through the input owners' observed identities. Do not copy the selection-research eight-file collector unchanged or hide Git failures as clean/empty evidence. Shared hashing follows #205; data collection stays #191/#206. No strict-mode switch, software pin, numerical threshold, GPU/Colab campaign, vendor fetch or archived-model relocation is implied.

**Comment by magilliam27** [comment]
## Cross-link: #111 is the same shape at the trainer layer

#111 (`Trainer reports -inf co-metrics, and co-metrics need not match the saved checkpoint`) states the same defect as this ticket one layer in: **the artifact does not describe the run that produced it.**

This ticket covers environment and code identity. #111 covers the reported co-metrics: `best_val_ic` and `best_val_rank_ic` can be `-inf`, and even when finite they need not belong to the epoch whose checkpoint was actually saved.

Worth reading before starting here. Capturing environment and code identity faithfully, while the co-metrics shipped alongside them can belong to a different epoch than the checkpoint, would leave "what produced this artifact?" unanswerable — with more metadata attached, which is worse than an obvious gap.

Recorded as a cross-link rather than a parent/child edge; the reasoning is on #111. #111 remains open.

**Comment by magilliam27** [comment]
PM recomputed native prerequisites after the owner-approved PR #204 merge: #142 is closed, so the documentation gate is satisfied. Live issue/comments/task/worktree and all eight open-PR inventories found no existing #144 owner. One Astra / Extra High isolated worker is reserved and assigned before branching; exact paths and execution-time acceptance refinements are now in the body. #191/#208 share metadata integration, so their data/attachment work is excluded and shared hunks must be coordinated before edits. The first proposed public capture-to-read-back boundary is not owner-confirmed; the current single pending owner question remains #205's reader seam. New worker begins bounded preparation and routes its own first seam to PM without a parallel question. This is a claim/dispatch, not implementation completion.

**Comment by magilliam27** [comment]
Dispatch correction: no #144 worker was created. PM is awaiting direct owner authorization for the new task, so the temporary assignment is released and the body now labels the path list as a proposal rather than an active claim. The merged PR #204/#142 gate remains satisfied; no implementation, test or worker startup is claimed. Earlier reservation wording is superseded. The read-only execution-time provenance packet remains available for the eventual owner.

**Comment by magilliam27** [comment]
Direct owner authorization now received in PM task: "yes proceed" to starting one Astra / Extra High worker for this ticket. Live recheck found no existing owner/task and only closed #142 as its native prerequisite; main remains f81573a. The issue is reclaimed before task creation, its path list restored as the bounded ownership claim, and the dispatch row updated. The previous rejected attempt created no worker, so this does not duplicate a task. New worker will coordinate shared hunks, prepare its first public seam for owner confirmation through PM, and use the strict Pocock workflow. This task-start approval is not a merge, compute authorization or blanket test-seam approval.

**Comment by magilliam27** [comment]
Startup and bounded preparation verified for the owner-approved worker.

- Task: `01a098ee-6d58-7671-8094-24992dfc3a1c`.
- Isolated worktree: `C:/Users/magil/.codex/worktrees/39b9/MCI-GRU`.
- Branch: `codex/144-execution-provenance`, created after the claim from freshly fetched main `f81573a14a35cdc53975640c28df8d3fc39b5756`.
- PR #204 is merged; #142 is closed. The eight open PRs contain no duplicate implementation for this ticket. Source/test/registry diff remains empty.

The PM supplied the existing read-only handoff from continuity notes; no standalone packet path is confirmed. Its narrow findings match the inspected current call sites: the eight-file selection-research hash collector is insufficient for training source recovery; Git failures currently become empty bytes; the redacted resolved-config artifact precedes preparation; actual seeds/device/AMP belong at their execution sites.

**One proposed public test boundary — pending owner confirmation through PM:** execution capture to persisted read-back in the new `mci_gru/evaluation/execution_provenance.py`. Tentative operations are `capture_execution_start(...)` and `read_execution_provenance(record_reference)`; exact names/schema remain engineering choices.

Capture identifies an attempt/window before preparation and retains effective redacted config bytes/digest, recoverable relevant source bytes (including dirty tracked and relevant untracked source), and timestamped observed environment state. Read-back uses only the retained evidence. Missing/altered artifacts and unsupported formats fail explicitly; failed Git or unavailable environment observations remain explicit failures/unknowns. Credentials are excluded. Existing input-path redactions are subsequently bound through the input owners' observed identities; this module does not collect data fingerprints.

**First vertical proof at this boundary:** use a tiny fixture repository with known tracked training-source bytes, a dirty edit, and an untracked training module; capture a known effective config and execution-time environment. Change the live checkout and config objects, then disable live observations. Public read-back must recover the original retained bytes/config/digest and original environment observations. Without a completion event, the attempt remains explicitly incomplete. Follow-on cases at this same boundary will remove or alter retained evidence and exercise Git failures, one red/green slice at a time. This establishes recovery of recorded state, not bitwise training equality or restoration of a running process.

PM is coordinating the narrow `run_experiment.py` / `experiment_summary.py` linkage hunks with #191/#208. #191 retains observed inputs and legacy `data_file_*`; #206 retains provider snapshots; #205 retains the shared file-hashing extraction; #208 retains attachment/bundle behavior. No shared source edits precede that coordination.

The owner has not approved this seam. #205's manifest-reader question remains first in the PM queue. No tests, implementation, registry regeneration, commit, PR, or training run has started; #144 remains open.

**Comment by magilliam27** [comment]
The existing worker has resumed under successor PM coordination; the stale dispatch paragraph has been updated without changing ownership or approvals. Its branch is clean at main `f81573a14a35cdc53975640c28df8d3fc39b5756`. The live native dependency list contains only closed #142.

The same first capture-to-persisted-read-back boundary in [comment 5651056929](https://github.com/magilliam27/MCI-GRU/issues/144#issuecomment-5651056929) remains awaiting explicit owner confirmation through PM. The earlier queue position behind #205 is historical. A restart is not boundary approval. The recommendation and isolated-module design have been returned to PM; no tests or implementation have started.

Shared metadata edits remain pending #191 acknowledgment; #208's ownership split is retained. Future trainer reporting must coordinate with #111, and registry regeneration with the separate unmerged PR #217. Later member-execution and stock/index/walk-forward boundaries remain unapproved. This is preparation/coordination status, not implementation completion.

**Comment by magilliam27** [comment]
Owner decision recorded: the owner answered “Approve this first step (recommended)” to the proposed execution-record capture and retained-evidence readback behavior.

Approved first boundary: before data preparation, retain an attempt/window record of actual relevant source bytes, including dirty tracked edits and relevant untracked training code; effective settings with secrets removed and their digest; and timestamped software-environment observations. Later inspection must use only retained evidence. Missing, altered or unsupported retained artifacts produce explicit errors; unavailable observations remain explicit failures or unknowns. Without completion evidence, the attempt stays incomplete. Existing input-identity owners retain data fingerprints.

First proof: capture a small repository fixture, configuration and environment observation; change the live source/settings and disable live observation; verify public readback recovers the original retained record. Exercise corruption and unavailable-observation behavior through subsequent individual red/green slices at this same boundary. This proves recorded-evidence recovery, not training equivalence.

This releases the existing worker for that first boundary and its already-declared owned scope. #191 shared-metadata coordination is acknowledged in the existing handoff. Actual seed/device/precision reporting and stock/index/window integration remain later, unapproved boundaries. No merge or additional compute approval is inferred.

**Comment by magilliam27** [comment]
The owner explicitly approved publication of exact first-slice commit `f3b931671e1ab12d047f3951d4851fc8046b6abe`. It is now published in [partial draft PR #221](https://github.com/magilliam27/MCI-GRU/pull/221). This covers only the approved execution-start capture/readback boundary and does not complete this issue. A registry-only conflict against main `2030f91b` prevents final CI. The existing owner is assigned routine local integration and regenerated combined evidence; subsequent publication will be tracked separately. Actual member seed/device/precision, training integration, completion and bundle attachment remain later boundaries. No main merge is authorized.

**Comment by magilliam27** [comment]
Owner-approved registry integration `90916f99737d232b71a3619925d7b712b7137856` is published to partial draft PR #221. [Exact-head CI 35484858805](https://github.com/magilliam27/MCI-GRU/actions/runs/35484858805) passed lint, docs/registry checks, capability-independent tests and end-to-end smoke. PR is mergeable against main `2030f91b`; the capture/readback module and focused tests remain unchanged from the approved first slice. The broader issue, later runtime/member boundaries and main merge remain separate.
